pkg/maps/egressmap/egress.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package egressmap
+
+import (
+	"fmt"
+	"net"
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/types"
+)
+
+const (
+	MapName    = "cilium_egress_v4"
+	MaxEntries = 16384
+)
+
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type Key4 struct {
+	PrefixLen uint32
+	SourceIP  types.IPv4
+	DestCIDR  types.IPv4
+}
+
+func (k *Key4) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+func (k *Key4) NewValue() bpf.MapValue    { return &EgressInfo4{} }
+func (k *Key4) String() string            { return fmt.Sprintf("%s %s %d", k.SourceIP, k.DestCIDR, k.PrefixLen) }
+
+func getStaticPrefixBits() uint32 {
+	staticMatchSize := unsafe.Sizeof(Key4{})
+	staticMatchSize -= unsafe.Sizeof(Key4{}.PrefixLen)
+	staticMatchSize -= unsafe.Sizeof(Key4{}.DestCIDR)
+	return uint32(staticMatchSize) * 8
+}
+
+// NewKey returns a new Key4 instance
+func NewKey(src, dst net.IP, mask net.IPMask) Key4 {
+	result := Key4{}
+
+	ones, _ := mask.Size()
+
+	copy(result.SourceIP[:], src.To4())
+	copy(result.DestCIDR[:], dst.To4())
+	result.PrefixLen = getStaticPrefixBits() + uint32(ones)
```

Comment:
I think there's something wrong here with the prefix length:
```
$ cilium bpf egress update 10.0.0.69 0.0.0.0/0 10.0.1.15 192.168.36.12
$ cilium bpf egress update 10.0.0.69 8.0.0.0/8 10.0.0.3 192.168.36.11
$ cilium bpf egress ls
key                    value
10.0.0.69 8.0.0.0 40   10.0.0.3 192.168.36.11    
10.0.0.69 0.0.0.0 32   10.0.1.15 192.168.36.12
```

Reply:
The prefix is actually 32 + the next cidr, because the first ip needs to be fully matched.

Reply:
Ah, right. Makes sense.

Could we remove 32 when we display then?